Lib/fractions.py

```diff
@@ -380,32 +380,139 @@ def reverse(b, a):
+    # denominator d could be much larger than g.
+    #
+    # We should special-case g == 1 (and g2 == 1), since 60.8% of
+    # randomly-chosen integers are coprime:
+    # https://en.wikipedia.org/wiki/Coprime_integers#Probability_of_coprimality
+    # Note, that g2 == 1 always for fractions, obtained from floats: here
+    # g is a power of 2 and the unnormalized numerator t is an odd integer.
+    #
+    # 2) Consider multiplication
+    #
+    # Let g1 = gcd(na, db) and g2 = gcd(nb, da), then
+    #
+    #            na*nb    na*nb    (na//g1)*(nb//g2)
+    #     a*b == ----- == ----- == -----------------
+    #            da*db    db*da    (db//g1)*(da//g2)
+    #
+    # Note, that after divisions we're multiplying smaller integers.
+    #
+    # Also, the resulting fraction is normalized, because each of
+    # two factors in the numerator is coprime to each of the two factors
+    # in the denominator.
+    #
+    # Indeed, pick (na//g1).  It's coprime with (da//g2), because input
+    # fractions are normalized.  It's also coprime with (db//g1), because
+    # common factors are removed by g1 == gcd(na, db).
+    #
+    # As for addition/subtraction, we should special-case g1 == 1
+    # and g2 == 1 for same reason.  That happens also for multiplying
+    # rationals, obtained from floats.
+
     def _add(a, b):
         """a + b"""
-        da, db = a.denominator, b.denominator
-        return Fraction(a.numerator * db + b.numerator * da,
-                        da * db)
+        na, da = a.numerator, a.denominator
+        nb, db = b.numerator, b.denominator
+        g = math.gcd(da, db)
+        if g == 1:
+            return Fraction(na * db + da * nb, da * db, _normalize=False)
+        s = da // g
+        t = na * (db // g) + nb * s
+        g2 = math.gcd(t, g)
+        if g2 == 1:
+            return Fraction(t, s * db, _normalize=False)
+        return Fraction(t // g2, s * (db // g2), _normalize=False)
 
     __add__, __radd__ = _operator_fallbacks(_add, operator.add)
 
     def _sub(a, b):
         """a - b"""
-        da, db = a.denominator, b.denominator
-        return Fraction(a.numerator * db - b.numerator * da,
-                        da * db)
+        na, da = a.numerator, a.denominator
+        nb, db = b.numerator, b.denominator
+        g = math.gcd(da, db)
+        if g == 1:
+            return Fraction(na * db - da * nb, da * db, _normalize=False)
+        s = da // g
+        t = na * (db // g) - nb * s
+        g2 = math.gcd(t, g)
+        if g2 == 1:
+            return Fraction(t, s * db, _normalize=False)
+        return Fraction(t // g2, s * (db // g2), _normalize=False)
 
     __sub__, __rsub__ = _operator_fallbacks(_sub, operator.sub)
 
     def _mul(a, b):
         """a * b"""
-        return Fraction(a.numerator * b.numerator, a.denominator * b.denominator)
+        na, da = a.numerator, a.denominator
+        nb, db = b.numerator, b.denominator
+        g1 = math.gcd(na, db)
+        if g1 > 1:
```

Comment:
I think the "> 1" tests should be removed.  The cost of a test is in the same ballpark as the cost of a division.  The code is clearer if you just always divide.  The performance gain mostly comes from two small gcd's being faster than a single big gcd.

Reply:
>    I think the "> 1" tests should be removed.

Maybe, I didn't benchmark this closely, but...

>    The cost of a test is in the  same ballpark as the cost of a division.

This is a micro-optimization, yes.  But ">" has not same cost as "//" even in the CPython:
```
$ ./python -m timeit -s 'a = 1' -s 'b = 1' 'a > b'
5000000 loops, best of 5: 82.3 nsec per loop
$ ./python -m timeit -s 'a = 111111111111111111111111111111111' -s 'b = 1' 'a > b'
5000000 loops, best of 5: 86.1 nsec per loop
$ ./python -m timeit -s 'a = 11111' -s 'b = 1' 'a // b'
2000000 loops, best of 5: 120 nsec per loop
$ ./python -m timeit -s 'a = 111111111111111111111111' -s 'b = 1' 'a // b'
1000000 loops, best of 5: 203 nsec per loop
$ ./python -m timeit -s 'a = 1111111111111111111111111111111111111111111111' -s 'b = 1' 'a // b'
1000000 loops, best of 5: 251 nsec per loop
$ ./python -m timeit -s 'a = 111111111111111111111111111111122222222222222111111111111111' \
 -s 'b = 1' 'a // b'
1000000 loops, best of 5: 294 nsec per loop
```
Maybe my measurements are primitive, but there is some pattern...

> The code is clearer if you just always divide.

I tried to explain reasons for branching in the comment above code.  Careful reader might ask: why you don't include same optimizations in the `_mul` as in `_add/sub`?

On another hand, gmplib doesn't use thise optimizations (c.f. same in mpz_add/sub).  SymPy's PythonRational class - too.

So, I did my best in defending those branches and I'm fine with removing, if @tim-one has no arguments against.

>    The performance gain mostly comes from two small gcd's being faster than a single big gcd.

This is more important, yes.

Reply:
> The cost of a test is in the same ballpark as the cost of a division.

? Comparing `g` to 1 takes, at worst, small constant time, independent of how many digits `g` has. But `n // 1` takes time proportional to the number of internal digits in `n` - it's a slow, indirect way of making a physical copy of `n`. Since we expect `g` to be 1 most of the time, and these are unbounded ints, the case for checking `g > 1` seems clear to me.

Reply:
I don't mind either way.

If it were just me, I would leave out the '> 1' tests because:
1) For [small inputs](https://bugs.python.org/issue43420#msg388215), the code is about 10% faster without the guards.
2) For random 100,000 bit numerators and denominators where the gcd's were equal to 1, my timings show no discernible benefit.
3) The code and tests are simpler without the guards.

Reply:
That just doesn't make much sense to me. On my box just now, with the released 3.9.1:

> C:\Windows\System32>py -m timeit -s "g = 1" "g > 1"
> 10000000 loops, best of 5: 24.2 nsec per loop
>
> C:\Windows\System32>py -m timeit -s "g = 1; t = 1 << 100000" "g > 1; t // g"
> 10000 loops, best of 5: 32.4 usec per loop

So doing the useless division not only drove the per-iteration measure from 24.2 to 32.4, the _unit_ increased by a factor of 1000(!), from nanoseconds to microseconds. That's using, by construction, an exactly 100,001 bit numerator.

And, no, removing the test barely improves that relative disaster at all:

> C:\Windows\System32>py -m timeit -s "g = 1; t = 1 << 100000" "t // g"
> 10000 loops, best of 5: 32.3 usec per loop

BTW, as expected, boost the number of numerator bits by a factor of 10 (to a million+1), and per-loop expense also becomes 10 times as much; cut it by a factor 10, and similarly the per-loop expense is also cut by a factor of 10.

With respect to the BPO message you linked to, it appeared atypical: _both_ gcds were greater than 1: `(10 / 3)  * (6 / 5)`, so there was never any possibility of testing for `> 1` paying off (while in real life a gcd of 1 is probably most common, and when multiplying Fractions obtained from floats gcd=1 is certain unless one of the inputs is 0).